gadgets/traceloop/go/program.go

```diff
@@ -0,0 +1,649 @@
+				typ:                sysEvent.Typ,
+				contNr:             sysEvent.ContNr,
+				cpu:                sysEvent.Cpu,
+				id:                 sysEvent.Id,
+				pid:                sysEvent.Pid,
+				comm:               fromCString(sysEvent.Comm[:]),
+				mountNsID:          reader.mntnsID,
+			}
+
+			var typeMap *map[uint64][]*syscallEvent
+			switch event.typ {
+			case syscallEventTypeEnter:
+				event.args = make([]uint64, syscallArgs)
+				for i := uint8(0); i < syscallArgs; i++ {
+					event.args[i] = sysEvent.Args[i]
+				}
+
+				typeMap = &syscallEnterEventsMap
+			case syscallEventTypeExit:
+				event.retval = sysEvent.Args[0]
+
+				typeMap = &syscallExitEventsMap
+			default:
+				// Rather than returning an error, we skip this event.
+				api.Debugf("type %d is not a valid type for syscallEvent, received data are: %v", event.typ, record)
+
+				continue
+			}
+
+			if _, ok := (*typeMap)[event.monotonicTimestamp]; !ok {
+				(*typeMap)[event.monotonicTimestamp] = make([]*syscallEvent, 0)
+			}
+
+			(*typeMap)[event.monotonicTimestamp] = append((*typeMap)[event.monotonicTimestamp], event)
+		case alignSize(unsafe.Sizeof(*sysEventCont)):
+			sysEventCont = (*traceloopSyscallEventContT)(unsafe.Pointer(&record[0]))
+
+			event := &syscallEventContinued{
+				monotonicTimestamp: sysEventCont.MonotonicTimestamp,
+				index:              sysEventCont.Index,
+			}
+
+			if sysEventCont.Failed != 0 {
+				event.param = "(Failed to dereference pointer)"
+			} else if sysEventCont.Length == useNullByteLength {
+				// 0 byte at [C.PARAM_LENGTH - 1] is enforced in BPF code
+				event.param = fromCString(sysEventCont.Param[:])
+			} else {
+				event.param = fromCStringN(sysEventCont.Param[:], int(sysEventCont.Length))
+			}
+
+			// Remove all non unicode character from the string.
+			event.param = strconv.Quote(event.param)
+
+			_, ok := syscallContinuedEventsMap[event.monotonicTimestamp]
+			if !ok {
+				// Just create a 0 elements slice for the moment, the ContNr will be
+				// checked later.
+				syscallContinuedEventsMap[event.monotonicTimestamp] = make([]*syscallEventContinued, 0)
+			}
+
+			syscallContinuedEventsMap[event.monotonicTimestamp] = append(syscallContinuedEventsMap[event.monotonicTimestamp], event)
+		default:
+			api.Debugf("size %d does not correspond to any expected element, which are %d and %d; received data are: %v", size, unsafe.Sizeof(sysEvent), unsafe.Sizeof(sysEventCont), record)
+		}
+	}
+
+	// Let's try to publish the events we gathered.
+	for enterTimestamp, enterTimestampEvents := range syscallEnterEventsMap {
+		for _, enterEvent := range enterTimestampEvents {
+			syscallName, err := api.GetSyscallName(enterEvent.id)
+			if err != nil {
+				return nil, fmt.Errorf("getting syscall name: %w", err)
+			}
+
+			event := &tracelooptypes.Event{
+				Event: eventtypes.Event{
+					Type:      eventtypes.NORMAL,
+					Timestamp: timestampFromEvent(enterEvent),
```

Comment:
Any way to use the timestamp formatter instead of having to handle it again here?

Reply:
If [`gadget_timestamp`](https://www.inspektor-gadget.io/docs/v0.35.0/gadget-devel/gadget-ebpf-api#gadget_timestamp) is used Inspektor Gadget will perform the uint to string conversion automatically. I this case the field isn't coming from ebpf, but I think it should work the same.

Reply:
`bpf_ktime_get_boot_ns()` was added in Linux 5.8 (released 4 years ago). Given that you are re-writing traceloop, I think it is fine to drop the compatibility with Linux < 5.8, and assume `bpf_ktime_get_boot_ns()` is available.